### PR TITLE
Conditionally spawn packet workers

### LINKED ISSUE #956: Improve code quality of unreceived-packets or unreceived-acks
Follow up from: https://github.com/informalsystems/ibc-rs/pull/907

The current factoring on the `unreceived-packets` and `unreceived-acks` CLIs rely on duplicate and code and might not use the best queries to get channel_id. Improve the code quality by providing a standard using a more efficient query route and method for getting the counter party chain id.

### PR DESCRIPTION
**Pull request by hu55a1n1**
Closes: #901 
Closes #956

## Description

Extracts `unreceived_packets()` and `unreceived_acknowledgements()` functions from `relayer_cli` and reuse in `spawn_workers_for_channel()` to spawn packet workers only when there are outstanding packets. 

______

For contributor use:

- [x] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [ ] If applicable: Unit tests written, added test to CI.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Updated relevant documentation (`docs/`) and code comments.
- [x] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: relayer/src/supervisor/spawn.rs [hunk 1]
```diff
@@ -514,22 +508,34 @@ impl<'a> SpawnContext<'a> {
...
-                )
-                .then(|| debug!("spawned Path worker: {}", path_object.short_name()));
+                self.workers
+                    .spawn(
+                        chain.clone(),
+                        counterparty_chain.clone(),
+                        &path_object,
+                        &self.config.read().expect("poisoned lock"),
+                    )
+                    .then(|| debug!("spawned Path worker: {}", path_object.short_name()));
```
Comment (adizere) [inline relayer/src/supervisor/spawn.rs:537]:
```suggestion
                    .then(|| debug!("spawned Packet worker: {}", path_object.short_name()));
```

**Review by adizere — COMMENTED** [review]
I did some quick tests and found that `hermes start` does not send out any pending packets. This may be because this branch is not even with latest master (and there was a bug we just recently fixed, which has to do with clearing pending packets).

To reproduce the behavior:

- start two chains with dev-env
- create a channel
- `hermes tx raw ft-transfer ibc-1 ibc-0 transfer channel-0 9999 -o 1000 -k user2`
- `hermes start` -- we see no activity, so the packet remains unsend

I suggest to first merge master into this branch and then maybe that will fix the problem.

**Comment by hu55a1n1** [comment]
Hi @adizere! Thanks for the suggestions! 💐 I just rebased onto master and I can confirm that the pending packets issue is re
solved ->
```bash
$ hermes start
# ....
Jul 26 19:11:17.156  INFO [ibc-1:transfer/channel-0 -> ibc-0] clearing pending packets height=Some(Height { revision: 1, height: 283 })
Jul 26 19:11:17.177  INFO [ibc-1:transfer/channel-0 -> ibc-0] finished scheduling pending packets clearing height=Some(Height { revision: 1, height: 283 })
```

File: relayer/src/supervisor/spawn.rs [hunk 2]
```diff
@@ -514,22 +508,34 @@ impl<'a> SpawnContext<'a> {
...
+                    .is_empty();
+
+            if outstanding_packets {
+                // create the Packet object and spawn worker
+                let path_object = Object::Packet(Packet {
+                    dst_chain_id: counterparty_chain.id(),
+                    src_chain_id: chain.id(),
+                    src_channel_id: channel.channel_id,
+                    src_port_id: channel.port_id,
+                });
```
Comment (adizere) [inline relayer/src/supervisor/spawn.rs:528]: The `Packet` object we're creating (line 523) is for the relaying path from `chain` (source chain) to `counterparty_chain` (destination chain). The worker for this object will be responsible for relaying packets based on events generated by source  `chain` and having destination `counterparty_chain`.

This worker will handle packets flowing strictly for the direction `chain` -> `counterparty_chain`.

The problem is that the logic in `outstanding_packets` algorithm (line 511-519) is not consistent with the direction of the worker. More specifically, the direction computed by `outstanding_packets` is reversed. When we compute `outstanding_packets`, we are considering the two following cases:

1. `unreceived_packets` for the direction `counterparty_chain` -> `chain`
2. `unreceived_acknowledgements` for the direction `counterparty_chain` -> `chain`.

To fix the problem, we need to check for unreceived packets and unreceived acknowledgements in the opposite direction, for packets on path `chain` -> `counterparty_chain`. I am leaving a suggestion for a patch below:

```suggestion
            let outstanding_packets =
                !unreceived_packets(counterparty_chain.as_ref(), chain.as_ref(), channel.clone())?
                    .is_empty()
                    || !unreceived_acknowledgements(
                        counterparty_chain.as_ref(),
                        chain.as_ref(),
                        channel.clone(),
                    )?
                    .is_empty();

            if outstanding_packets {
                // create the Packet object and spawn worker
                let path_object = Object::Packet(Packet {
                    dst_chain_id: counterparty_chain.id(),
                    src_chain_id: chain.id(),
                    src_channel_id: channel.channel_id,
                    src_port_id: channel.port_id,
                });
```

**Review by adizere — COMMENTED** [review]
This is almost ready, except for a corner-case highlighted in a comment.

To test for this corner-case and see the bug in action, use the following:

1. start two chains ibc-0 and ibc-1
2. send two packets on the path `ibc-0 -> ibc-1`
    - `hermes tx raw ft-transfer ibc-1 ibc-0 transfer channel-0 9999 -o 1000 -n 2 -k user2`
3. test that the packets are unreceived on the destination `ibc-1`
    - `hermes query packet unreceived-packets ibc-1 transfer channel-0`
    - you should see something as follows (possibly different sequence numbers)
```
Success: [
    2,
    3,
]
```

4. `hermes start` will not clear the two packets
    - a worker will start, but for the path `ibc-1->ibc-0`, and not for the correct path (which should be `ibc-0 -> ibc-1`)

```
Jul 29 10:47:34.373 DEBUG spawning worker for object packet::channel-0/transfer:ibc-1->ibc-0
Jul 29 10:47:34.373 DEBUG spawned Packet worker: packet::channel-0/transfer:ibc-1->ibc-0
Jul 29 10:47:34.373 DEBUG done spawning workers for chain ibc-1 and channel channel-0
```

File: relayer/src/supervisor/spawn.rs (hunk 2 above)
Comment (hu55a1n1) [reply relayer/src/supervisor/spawn.rs]: Thanks for noticing and helping fix this bug @adi! That's a good catch! 👌 
Sorry for not testing this well enough and relying only on the E2E tests. Do you think we should create an issue to add a test for this in `run.py`?

Comment (adizere) [reply relayer/src/supervisor/spawn.rs]: I don't actually understand why the E2E tests did not catch this. It's possible that it's because there were packets pending on _both_ directions of a path, and then workers were spawned for both directions, which masked the problem.

> Do you think we should create an issue to add a test for this in run.py?

Your suggestion sounds very good! But it would be good to confirm what is the reason why run.py was unable to catch this corner-case in the first place, then we can translate this into an issue.

Comment (hu55a1n1) [reply relayer/src/supervisor/spawn.rs]: I just tested locally and I can confirm that the original bug has been fixed. 👍 Still trying to figure out why the E2E tests didn't catch it.

Comment (hu55a1n1) [reply relayer/src/supervisor/spawn.rs]: So the E2E tests do fail one commit earlier even on multiple retries (at least locally) -> 
```python
Traceback (most recent call last):
  File "/relayer/e2e/run.py", line 276, in <module>
    main()
  File "/relayer/e2e/run.py", line 247, in main
    passive_packets(config, ibc0, ibc1, port_id, ibc0_chan_id, ibc1_chan_id)
  File "/relayer/e2e/run.py", line 98, in passive_packets
    assert (len(unreceived) == 0), (unreceived,
AssertionError: ([5, 6, 7], 'unreceived acks mismatch (expected 0)')
```

I think this happened before as well, but we ignored it thinking it was caused by #1216 (that was fixed recently). Still investigating the logs...

Comment (hu55a1n1) [reply relayer/src/supervisor/spawn.rs]: I did some more testing and found that the `unreceived_packets()` function expects the channel on the party's side as it's 3rd arg ->
```rust
pub fn unreceived_packets(
    chain: &dyn ChainHandle,
    counterparty_chain: &dyn ChainHandle,
    channel: IdentifiedChannelEnd,
) -> Result<Vec<u64>, Error>;
```
But when we're calling it from `spawn_workers_for_channel()` ->
```rust
unreceived_packets(counterparty_chain.as_ref(), chain.as_ref(), channel.clone())?;
```
We're giving it the channel on the other chain. So we should've instead given it the `counterparty_channel` as the 3rd arg. Just fixed it.

File: relayer/src/supervisor/spawn.rs [hunk 3]
```diff
@@ -508,15 +508,29 @@ impl<'a> SpawnContext<'a> {
...
+                chain.as_ref(),
+                counterparty_channel.clone(),
+            )?
+            .is_empty()
+                || !unreceived_acknowledgements(
+                    counterparty_chain.as_ref(),
+                    chain.as_ref(),
+                    counterparty_channel.clone(),
+                )?
+                .is_empty();
```
Comment (romac) [inline relayer/src/supervisor/spawn.rs]: This code is not particularly easy to read because of all the nested function calls. Can we add some `let` bindings to give names to the sub-expressions?

File: relayer/src/supervisor/spawn.rs [hunk 4]
```diff
@@ -514,22 +508,48 @@ impl<'a> SpawnContext<'a> {
...
```
Comment (romac) [inline relayer/src/supervisor/spawn.rs:533]: I find that this code is not particularly easy to read because of all the nested function calls. Can we add some `let` bindings to give names to the sub-expressions?

File: relayer/src/supervisor/spawn.rs [hunk 5]
```diff
@@ -514,22 +508,48 @@ impl<'a> SpawnContext<'a> {
...
+            let counterparty_channel = IdentifiedChannelEnd::new(
+                channel.channel_end.counterparty().port_id.clone(),
+                channel
+                    .channel_end
+                    .counterparty()
+                    .channel_id
+                    .clone()
+                    .unwrap(),
+                counterparty_channel.unwrap(), // safe because chan_state_dst is Open
+            );
```
Comment (ancazamfir) [inline relayer/src/supervisor/spawn.rs:520]: Wondering if we should change the `channel_on_destination()` to return `Option<IdentifiedChannelEnd>` instead of rebuilding it here.

**Review by ancazamfir — APPROVED** [review]
Looks good, thanks @hu55a1n1!
Just a note that with the changes in this PR the packet worker metric will show the number of packet workers that did meaningful work since the start of the current hermes instance (vs. the number of open channels x 2)

File: relayer/src/supervisor/spawn.rs (hunk 4 above)
Comment (adizere) [reply relayer/src/supervisor/spawn.rs]: 850ca46d

File: relayer/src/supervisor/spawn.rs (hunk 5 above)
Comment (adizere) [reply relayer/src/supervisor/spawn.rs]: Good idea, I like it! did it here 9a3b4f2c

**Comment by adizere** [comment]
@hu55a1n1 can you please double-check on the last two commits (850ca46, 9a3b4f2) that my modifications are sound? I think we should merge this today.

File: relayer/src/supervisor/spawn.rs [hunk 6]
```diff
@@ -513,22 +507,41 @@ impl<'a> SpawnContext<'a> {
...
+            let has_acks = !unreceived_acknowledgements(
+                counterparty_chain.as_ref(),
+                chain.as_ref(),
+                counterparty_channel,
+            )?
+            .is_empty();
+
+            // If there are any outstanding packets or acks to send, spawn the worker
+            if has_packets || has_acks {
+                // create the Packet object and spawn worker
```
Comment (hu55a1n1) [inline relayer/src/supervisor/spawn.rs:528]: I think it is better to extract the `unreceived_*` functions into closure so as to benefit from short-circuiting on the `if` on line 527. 
```suggestion
            let has_packets = || -> bool {
                !unreceived_packets(
                    counterparty_chain.as_ref(),
                    chain.as_ref(),
                    counterparty_channel.clone(),
                )
                .unwrap_or_default()
                .is_empty()
            };

            let has_acks = || -> bool {
                !unreceived_acknowledgements(
                    counterparty_chain.as_ref(),
                    chain.as_ref(),
                    counterparty_channel.clone(),
                )
                .unwrap_or_default()
                .is_empty()
            };

            // If there are any outstanding packets or acks to send, spawn the worker
            if has_packets() || has_acks() {
                // create the Packet object and spawn worker
```

**Comment by hu55a1n1** [comment]
> @hu55a1n1 can you please double-check on the last two commits ([850ca46](https://github.com/informalsystems/ibc-rs/commit/850ca46de5e9460d13fb563095d30e50c6b7eb10), [9a3b4f2](https://github.com/informalsystems/ibc-rs/commit/9a3b4f2c6ae546a5a6b5ca8d3aae0f75bcc50ff9)) that my modifications are sound? I think we should merge this today.

Thanks for the help @adizere! I just added a suggestion to use closures for achieving lazy evaluation with better readability. Otherwise, the PR looks like it's ready to be merged. 👍

**Review by adizere — APPROVED** [review]
LGTM. Great work Shoaib! Thank you for your patience with this PR, it was a bit more tricky one!
